src/Database/LSMTree/Internal.hs

```diff
@@ -1369,42 +1371,149 @@ unions ts = do
-    -- completely fresh. It will have an empty write buffer and no runs
-    -- in the normal levels. All the existing runs get squashed down into
-    -- a single run before rejoining as a last level.
-    let conf = tableConfig (NE.head ts)
-
     -- We acquire a read-lock on the session open-state to prevent races, see
     -- 'sessionOpenTables'.
     modifyWithActionRegistry
       (atomically $ RW.unsafeAcquireReadAccess (sessionState sesh))
-      (\_ -> atomically $ RW.unsafeReleaseReadAccess (sessionState sesh)) $ \reg -> \case
+      (\_ -> atomically $ RW.unsafeReleaseReadAccess (sessionState sesh)) $
+      \reg -> \case
         SessionClosed -> throwIO ErrSessionClosed
         seshState@(SessionOpen seshEnv) -> do
-          contents <-
-            forM ts $ \t -> do
-              withOpenTable t $ \tEnv ->
-                -- The table contents escape the read access, but we just added references
-                -- to each run so it is safe.
-                RW.withReadAccess (tableContent tEnv) (duplicateTableContent reg)
-
-          content <-
-            error "unions: combine contents into merging tree" $ -- TODO
-              contents
-
-          t <-
-            newWith
-              reg
-              sesh
-              seshEnv
-              conf
-              (error "unions: ArenaManager") -- TODO
-              content
-
+          t <- unionsInOpenSession sesh seshEnv reg ts
           pure (seshState, t)
 
+{-# SPECIALISE unionsInOpenSession ::
+     Session IO h
+  -> SessionEnv IO h
+  -> ActionRegistry IO
+  -> NonEmpty (Table IO h)
+  -> IO (Table IO h) #-}
+unionsInOpenSession ::
+     (MonadSTM m, MonadMask m, MonadMVar m, MonadST m)
+  => Session m h
+  -> SessionEnv m h
+  -> ActionRegistry m
+  -> NonEmpty (Table m h)
+  -> m (Table m h)
+unionsInOpenSession sesh seshEnv reg ts = do
+
+    -- The TableConfig for the new table is taken from the first
+    -- table in the union. This works because the new table is almost
+    -- completely fresh. It will have an empty write buffer and no
+    -- runs in the normal levels. All the existing runs get squashed
+    -- down into a single run before rejoining as a last level.
+    let conf = tableConfig (NE.head ts)
+
+    mts <- forM (NE.toList ts) $ \t ->
+      withOpenTable t $ \tEnv ->
+        RW.withReadAccess (tableContent tEnv) $ \tc -> do
+          -- tableContentToMergingTree duplicates all runs and merges
+          -- so the ones from the tableContent here do not escape
+          -- the read access.
+          mt <- withRollback reg
+                  (tableContentToMergingTree seshEnv conf reg tc)
+                  releaseRef
+          -- The mt here is a temporary value, since newPendingUnionMerge
+          -- will make its own references, so release mt at the end of
+          -- the action registry bracket
+          delayedCommit reg (releaseRef mt)
+          return mt
+    mt <- newPendingUnionMerge mts
```

Comment:
Do we need a `withRollBack` here? Can something fail in the remainder of the function, even just the tracing in `newWith`?

Reply:
done.